servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletContext.java

```diff
@@ -0,0 +1,591 @@
+        return null;
+    }
+
+    @Override
+    public URL getResource(String path) throws MalformedURLException {
+        return null;
+    }
+
+    @Override
+    public InputStream getResourceAsStream(String path) {
+        return null;
+    }
+
+    @Override
+    public String getRealPath(String path) {
+        return path;
+    }
+
+    @Override
+    public ServletRequestDispatcher getRequestDispatcher(String path) {
+        requireNonNull(path, "path");
+        final UrlMapper.Element<ServletRegistration> element = servletUrlMapper.getMappingObjectByUri(path);
+        final ServletRegistration servletRegistration = element.getObject();
+        if (servletRegistration == null) {
+            return null;
+        }
+
+        final ServletFilterChain filterChain = new ServletFilterChain(this, servletRegistration);
+        final List<FilterRegistration> filterList = filterChain.getFilterRegistrationList();
+        for (FilterRegistration registration : filterRegistrationMap.values()) {
+            filterList.add(registration);
+        }
+        filterUrlMapper.addMappingObjectsByUri(path,filterChain.getFilterRegistrationList());
+
+        final ServletRequestDispatcher dispatcher = new ServletRequestDispatcher(filterChain);
+        dispatcher.setMapperElement(element);
+        dispatcher.setPath(path);
+        return dispatcher;
+    }
+
+    @Override
+    public ServletRequestDispatcher getNamedDispatcher(String name) {
+        requireNonNull(name, "name");
+        final ServletRegistration servletRegistration = getServletRegistration(name);
+        if (servletRegistration == null) {
+            return null;
+        }
+
+        final ServletFilterChain filterChain = new ServletFilterChain(this, servletRegistration);
+        final List<FilterRegistration> filterList = filterChain.getFilterRegistrationList();
+        for (FilterRegistration registration : filterRegistrationMap.values()) {
+            for (String servletName : registration.getServletNameMappings()) {
+                if (servletName.equals(name)) {
+                    filterList.add(registration);
+                }
+            }
+        }
+
+        final ServletRequestDispatcher dispatcher = new ServletRequestDispatcher(filterChain);
+        dispatcher.setName(name);
+        return dispatcher;
+    }
+
+    @Override
+    public Servlet getServlet(String name) throws ServletException {
+        requireNonNull(name, "name");
+        final ServletRegistration registration = servletRegistrationMap.get(name);
+        if (registration == null) {
+            return null;
+        }
+        return registration.getServlet();
+    }
+
+    @Override
+    public Enumeration<Servlet> getServlets() {
+        final List<Servlet> list = new ArrayList<>();
+        for (ServletRegistration registration : servletRegistrationMap.values()) {
+            list.add(registration.getServlet());
+        }
+        return Collections.enumeration(list);
```

Comment:
We should make the list as immutable not just for `getServlets()`, but also for `getServletNames()`, getInitParameterNames(), getAttributeNames() etc.
```
return Collections.enumeration(servletRegistrationMap.values()
                                                     .stream()
                                                     .map(ServletRegistration::getServlet)
                                                     .collect(toImmutableList()));
```

Also, one thing we have to think about is that if the field is unmodifiable after `DefaultServletContext` is initialized, we can pass it when we create `DefaultServletContext`.
For example, initParamMap can't be changed after `DefaultServiceContext` is initialized.
So we can do something like:
```
class ServletBuilder {

    // Introduce a new class 
    private final DefaultServletContextBuilder servletContextBuilder;

    ServletBuilder(ServerBuilder serverBuiler, String contextPath) {
        servletContextBuilder = new DefaultServletContextBuilder(contextPath);
    }

    public ServletBuilder initParameter(String key, @Nullable String value) {
        requireNonNull(key, "key");
        servletContextBuilder.setInitParameter(key, value);
        return this;
    }

    public ServerBuilder build() {
        new DefaultServletService(servletContextBuilder.build());
        ...
    }
    ...
}

class ServletContext {

    ServletContext(String contextPath, Map<String, String> initParamMap ...) {
        this.initParamMap = ImmutableMap.copyOf(initParamMap);
        ...
    }
}
```

Reply:
Totally agree!

Reply:
> Also, one thing we have to think about is that if the field is unmodifiable after DefaultServletContext is initialized, we can pass it when we create DefaultServletContext.
> For example, initParamMap can't be changed after DefaultServiceContext is initialized

I will apply to ImmutableMap for initParamMap, other attribute will keep.

Reply:
What's the �pack()? Could you elaborate more on this please?

Reply:
pack() and unPack() method usually to encode, decode, complete header ... a object (example: ISO message) before send it. I think immutable and unImmutable may be like that.

Reply:
I understood your implementation by constructor but if we have many immutable fields, I think using constructor is difficult to implement.

Reply:
Ok, I got it. Then, could you pack other properties too so that they are not set after initialized?

Reply:
Yes, I think so :) .

Reply:
We should raise an exception in these methods if the context is initialized:
```
addServlet, addJspFile, addFilter, setSessionTrackingModes, addListener,
declareRoles, setSessionTimeout, setRequestCharacterEncoding, setResponseCharacterEncoding
```

Reply:
Agree, I will throw exception if the context is initialized